### ISSUE TITLE
Implement noise "modifiers" in the Animation Player

### ISSUE THREAD

**Issue by Arnklit** [godotengine/godot-proposals#6142]
### Describe the project you are working on

Various 3D games and add-ons

### Describe the problem or limitation you are having in your project

Adding animation noise or "filler animation" to an object, requires scripting or external tools. Some examples where this is useful; Animate a light source to move around randomly for a flame light source. Create simple animation for a character as an idle animation or talking animation.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Similar to how you can add noise modifiers to animation tracks in Blender, it would be very useful if you could do it in Godot's Animation Player.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

We don't have any concept similar to Blender's animation modifiers, do it would probably have to be a button you could enable on each track and click another button to open a pop-up with settings for the noise. 

![image](https://user-images.githubusercontent.com/4955051/213939779-b0a97300-d41a-4190-9023-5022785c8968.png)

Alternatively a "Noise Track" type could be added.

![image](https://user-images.githubusercontent.com/4955051/213939869-a0c862e1-acb4-476e-a21e-cc06e79c8b12.png)

Settings would be similar to what Blender's noise modifier has. The noise should also automatically loop correctly, if looping is set on the track. 

![image](https://user-images.githubusercontent.com/4955051/213939621-d41e67b0-0532-4c96-8b24-b2bec8754f02.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, the idea was to avoid scripting, so it was easy for animators / designers to add noise to animations.

### Is there a reason why this should be core and not an add-on in the asset library?

To integrate this well I think it would have to be in core.

**Comment by KoBeWi** [comment]
> No, the idea was to avoid scripting, so it was easy for animators / designers to add noise to animations.

They don't need to write the script. A programmer can write one and export relevant properties. So this is 100% doable as an addon.

**Comment by Arnklit** [comment]
> They don't need to write the script. A programmer can write one and export relevant properties. So this is 100% doable as an addon.

Totally fair. I just feel that this is something that would fit well into being integrated in the engine, so it's discoverable and easy to use.

I'm guessing it would be quite a bit less elegant as an addon as I don't think you can override or add stuff into the AnimationPlayer ui with an addon? So you would probably need to do something like adding an additional node you would need to add as a child of each object you wanted to add noise to.

**Comment by KoBeWi** [comment]
Animation resources from AnimationPlayer are easily editable. But procedural animation is easier to achieve using Tweens.

How would that modifications work btw? Would they be baked into animation? Randomized at the beginning (potentially different for each instance) or randomized every loop?

**Comment by Arnklit** [comment]
> How would that modifications work btw? Would they be baked into animation? Randomized at the beginning (potentially different for each instance) or randomized every loop?

Yeah I'm not sure on this, I would think they wouldn't be baked in, unless that was absolutely needed for performance reasons. Seems it would simply make the animations bloated in size for no good reason to bake them in.

I figured they would have to loop as the seed for the noise would the timeline for the animation, but there might be other ways of doing it.

I'm interested in trying to implement this myself, but figured it would be nice to get some feedback on the idea first.

**Comment by farfalk** [comment]
I know this is a bit out of scope, but i think a generalized and extendable "animation modifier" system for some type of tracks would be an interesting addition, with noise as a specific (and very useful) modifier.

**Comment by MrEliptik** [comment]
I agree with @farfalk, I think it would make the animationPlayer even more interesting. 

@KoBeWi suggtested using Tweens to do procedural animations but it's hard to know how it's going to look like and sometimes you only need one random value in a very complex predefined animation. 

I wanted to suggest having a way to add more random values or even variables in the animation player. I think the idea of having an animation modifier would be the best and could enable lots of different use cases.
